### PR TITLE
string_literal module to workaround memory leak caused by 'makeStringLiteral'

### PR DESCRIPTION
**Pull request by sebastianbaginski**
In order to fix this memory leak I needed to create a separate module
in plain C, just to prevent circular references to static variables
declared in `String` class.
What I mean is that it is (I think) impossible to fix this at ooc level,
as declaring the `string literal` arrays directly in `String` class causes
rock to generate references to `String__String__class()` when accessing them:
```ooc
String: class {
    _stringLiterals := static This* = null
    _stringLiteralsCount := static 0
...
}

makeStringLiteral: func (...) {
   // this line will cause `String__String__class()` access,
   // which will again call `makeStringLiteral` at some point
    if (String _stringLiteralsCount >= ...) {
        String _stringLiterals = realloc( ... ) // last realloc result always leaked
        ...
    }
}

```
This caused the `_stringLiterals` array to leak for some reason,
even if it was explicitly released at program exit.
I struggled to fix this at ooc level for some time, other idea was to use global variables
instead of class statics, but this is even worse as global variables are not initialized,
even when explicitly assigned:
```ooc
stringLiterals := static String* = null

String: class {...
```
generates following C code:
```C
// at file scope 'String.c'
static String__String* String__stringLiterals; // not initialized anywhere, contains garbage
...
```

In fact my first idea was just to use a static `VectorList` in `String` class, forget it, VectorList class init functions calls `makeStringLiteral` too.

So in the end I decided to write logically equavalent code but in plain C and
manage string literal memory in separate module referenced via extern functions. Of course it worked without problems.

tl;dr : I needed to create a C "plugin" for `String` in order to fix the leaked string literals

@fredrikbryntesson review
@marcusnaslund @fredrikbryntesson @thomasfanell  I know it would be the best to fix such things at ooc level, but believe me, I spent far too many hours trying to fix this problem in ooc. We already have hand-written C code for `Arrays`, so I think this one is not that bad. What do you think about this solution ?

### CODE REVIEW

File: source/system/String.ooc [hunk 1]
```diff
@@ -22,6 +28,8 @@ String: class {
 		_buffer = CharBuffer new(s, length)
 	}
 	free: override func {
+		if (this _buffer mallocAddr == 0)
```
Comment (sebastianbaginski) [inline source/system/String.ooc]: `buffer mallocAddr` is `0` only for string literals

**Comment by thomasfanell** [comment]
This is similar to something i've considered doing myself, so I think it's a good solution. Great job!

**Comment by sebastianbaginski** [comment]
> This is similar to something i've considered doing myself

Thanks, good to know that.  Have you also tried to workaround this in ooc ?

**Comment by thomasfanell** [comment]
When I had a look, I came to the conclusion that the string stuff was so deeply nested it wouldn't really be worth it. Judging what you wrote here, I think you came to the same conclusion.

File: source/system/string/string_literal.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef __STRINGLITERAL_FIX_H__
+#define __STRINGLITERAL_FIX_H__
+
```
Comment (fredrikbryntesson) [inline source/system/string/string_literal.h]: _FIX_ makes it feel like a temporary hack that should be changed very soon.

**Comment by fredrikbryntesson** [comment]
String.ooc and string_literal.c has a lot of empty lines.

**Comment by marcusnaslund** [comment]
Why not rather try to get rid of String all together?

**Comment by sebastianbaginski** [comment]
> Why not rather try to get rid of String all together?

So why do we have an issue addressing this memory leak specifically ?
Also, isn't `makeStringLiteral` hard-coded in the compiler ?

File: source/system/string/string_literal.h (hunk 2 above)
Comment (sebastianbaginski) [inline source/system/string/string_literal.h]: That was the idea, I'm happy that you thought so. This is a hack and should be removed as soon as we get rid of String class.

**Comment by fredrikbryntesson** [comment]
Yes, or the c-function is actually.

**Comment by sebastianbaginski** [comment]
@fredrikbryntesson newlines in .c removed
do you want me to remove them all from String.ooc in this PR ?

**Comment by sebastianbaginski** [comment]
OK it is done.
